tar/tar.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Copyright 2020 The FluxCD contributors. All rights reserved.
+// Adapted from: golang.org/x/build/internal/untar
+
+// Package tar provides ways to manage tarball files.
+package tar
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"time"
+
+	securejoin "github.com/cyphar/filepath-securejoin"
+)
+
+// defaultMaxUntarSize sets the default max amount of bytes that Untar will process.
+// Defaults to 100MB.
+const defaultMaxUntarSize = 100 << (10 * 2)
+
+type tarOpts struct {
+	// maxUntarSize represents the limit size (bytes) for archives being decompressed by Untar.
+	// When max is equal or less than 0, the size checks are disabled.
+	maxUntarSize int
+}
+
+// Untar reads the gzip-compressed tar file from r and writes it into dir.
+//
+// If dir is a relative path, it cannot ascend from the current working dir.
+// If dir exists, it must be a directory.
+func Untar(r io.Reader, dir string, inOpts ...TarOption) (err error) {
+	opts := tarOpts{
+		maxUntarSize: defaultMaxUntarSize,
```

Comment:
I don't think we should be enforcing a max size of 100MB, we'll break Flux for many users who have huge app repos.

Reply:
The suggestion here is for the default within the library. For the controllers the plan was to make this flag configurable, with no limit as default.

Reply:
How would you pass `no limit` to this function?

Reply:
Negative integer (or at least, that is what I proposed offline).

Reply:
Let's add a const for no limit.

Reply:
Added `tar.UnlimitedUntarSize` to represent the no limit value, although any negative integer is also acceptable.